dapp-core/src/main/kotlin/jp/co/soramitsu/dapp/listener/ReliableIrohaChainListener.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright D3 Ledger, Inc. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package jp.co.soramitsu.dapp.listener
+
+import com.rabbitmq.client.BuiltinExchangeType
+import com.rabbitmq.client.ConnectionFactory
+import com.rabbitmq.client.Delivery
+import io.reactivex.Observable
+import io.reactivex.subjects.PublishSubject
+import iroha.protocol.BlockOuterClass
+import mu.KLogging
+import java.io.Closeable
+import java.util.concurrent.ExecutorService
+
+private const val DEFAULT_LAST_READ_BLOCK = -1L
+
+/**
+ * Rabbit MQ implementation of Iroha chain listener
+ * @param rmqHost - rmq host name
+ * @param rmqExchange - rmq exchange name
+ * @param irohaQueue - name of queue to read Iroha blocks from
+ * @param consumerExecutorService - executor that is used to execure RabbitMQ consumer code.
+ */
+class ReliableIrohaChainListener(
+    private val rmqHost: String,
+    private val rmqPort: Int,
+    rmqExchange: String,
+    private val irohaQueue: String,
+    private val consumerExecutorService: ExecutorService
+) : Closeable {
+
+    private val factory = ConnectionFactory()
+
+    // Last read Iroha block number. Used to detect double read.
+    private var lastReadBlockNum: Long = DEFAULT_LAST_READ_BLOCK
+
+    private val conn by lazy {
+        factory.host = rmqHost
+        factory.port = rmqPort
+        factory.newConnection(consumerExecutorService)
+    }
+
+    private val channel by lazy { conn.createChannel() }
+
+    private var consumerTag: String? = null
+
+    private val observable: Observable<BlockOuterClass.Block>
+
+    init {
+        channel.exchangeDeclare(rmqExchange, BuiltinExchangeType.FANOUT, true)
+        channel.queueDeclare(irohaQueue, true, false, false, null)
+        channel.queueBind(irohaQueue, rmqExchange, "")
+        channel.basicQos(1)
+
+        val source = PublishSubject.create<BlockOuterClass.Block>()
+        val deliverCallback = { _: String, delivery: Delivery ->
+            // This code is executed inside consumerExecutorService
+            val block = iroha.protocol.BlockOuterClass.Block.parseFrom(delivery.body)
+            //TODO shall we ignore too old blocks?
+            if (ableToHandleBlock(block)) {
+                source.onNext(block)
+            } else {
+                logger.warn { "Not able to handle Iroha block ${block.blockV1.payload.height}" }
+            }
+        }
+        observable = source.map { block ->
+            logger.info { "New Iroha block from RMQ arrived. Height ${block.blockV1.payload.height}" }
+            block
+        }
+        consumerTag = channel.basicConsume(irohaQueue, true, deliverCallback, { _ -> })
```

Comment:
Blocks will be lost if no subscriber was set before calling `basicConsume`, which makes `ReliableIrohaChainListener` senseless.

Reply:
fixed